# 1. What is a lambda function in Python, and how does it differ from a regular function?

A lambda function in Python is a small anonymous function that can be defined without a name. It is created using the lambda keyword, followed by the function arguments and a colon, and then the expression to be evaluated. 

The result of the expression is automatically returned. The main difference between a lambda function and a regular function is that lambda functions are used for simple and concise operations, while regular functions are more suitable for complex tasks and reusable code.

# 2. Can a lambda function in Python have multiple arguments? If yes, how can you define and use them?

Yes, a lambda function in Python can have multiple arguments. You can define and use them by separating the arguments with commas after the 'lambda' keyword. 

For example, a lambda function with two arguments can be defined like this: 'lambda x, y: x + y'. To use the lambda function, you can call it and pass the required arguments like any other function.

# 3. How are lambda functions typically used in Python? Provide an example use case.

Lambda functions are typically used in Python when a small function is needed for a short and one-time operation, especially when passing a function as an argument to another function. 

They are commonly used with built-in functions like 'map()', 'filter()', and 'reduce()', which expect a function as one of their parameters. Lambda functions provide a convenient way to define these functions on the fly without the need for a separate named function.

# 4. What are the advantages and limitations of lambda functions compared to regular functions in python.

The advantages of lambda functions in Python are their conciseness and simplicity. They allow you to write small functions quickly and inline, reducing the need for defining separate named functions.

Lambda functions are also useful for functional programming paradigms and provide a more compact syntax. However, lambda functions have limitations. 

They can only contain a single expression, which means they cannot have multiple statements or complex logic. Additionally, lambda functions lack the self-documentation and reusability of regular functions.

# 5. Are lambda functions in Python able to access variables defined outside of their own scope Explain with an example.?

Yes, lambda functions in Python can access variables defined outside of their own scope. This is known as variable capture. Lambda functions can access variables from the surrounding scope in which they are defined, including global variables.

However, if the variable is mutable (e.g., a list or dictionary), modifying it inside the lambda function will have an effect on the variable outside the function. Here's an example:


In [6]:
def multiplier(n):
    return lambda x: x * n

double = multiplier(2)
print(double(5))  


10


In the above example, the lambda function double captures the variable n from its outer scope (multiplier function) and uses it in the expression x * n.



1.Lambda function to calculate the square of a given number:

In [7]:
square = lambda x: x ** 2
result = square(5)
print(result)  


25


Lambda function to find the maximum value in a list of integers:

In [8]:
numbers = [5, 3, 9, 2, 7]
maximum = max(numbers, key=lambda x: x)
print(maximum)  


9


Lambda function to filter out all the even numbers from a list of integers:

In [9]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
filtered = list(filter(lambda x: x % 2 == 0, numbers))
print(filtered)  


[2, 4, 6, 8, 10]


Lambda function to sort a list of strings in ascending order based on the length of each string:

In [10]:
strings = ['apple', 'banana', 'cherry', 'date', 'elderberry']
sorted_strings = sorted(strings, key=lambda x: len(x))
print(sorted_strings) 


['date', 'apple', 'banana', 'cherry', 'elderberry']


Lambda function that returns a new list containing the common elements between two lists:

In [11]:
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]
common_elements = list(filter(lambda x: x in list1, list2))
print(common_elements)  

[4, 5]


Recursive function to calculate the factorial of a given positive integer:

In [12]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)

result = factorial(5)
print(result) 


120


Recursive function to compute the nth Fibonacci number:

In [13]:
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)

result = fibonacci(6)
print(result)  


8


Recursive function to find the sum of all elements in a given list:

In [14]:
def sum_list(lst):
    if not lst:
        return 0
    else:
        return lst[0] + sum_list(lst[1:])

result = sum_list([1, 2, 3, 4, 5])
print(result)


15


Recursive function to determine whether a given string is a palindrome:

In [15]:
def is_palindrome(string):
    if len(string) <= 1:
        return True
    else:
        return string[0] == string[-1] and is_palindrome(string[1:-1])

result = is_palindrome("radar")
print(result)  


True


Recursive function to find the greatest common divisor (GCD) of two positive integers:

In [16]:
def gcd(a, b):
    if b == 0:
        return a
    else:
        return gcd(b, a % b)

result = gcd(24, 36)
print(result)  

12


# 6. Write a lambda function to calculate the square of a given number.

In [17]:
square = lambda x: x ** 2


In the above lambda function, x is the input parameter, and the expression x ** 2 calculates the square of x.

You can use this lambda function to calculate the square of a number by calling it and passing the desired number as an argument. Here's an example:

In [18]:
result = square(5)
print(result) 


25


n the example above, square(5) calls the lambda function with 5 as the input, and it returns the square of 5, which is 25.

# 7. Create a lambda function to find the maximum value in a list of integers.

In [19]:
numbers = [5, 3, 9, 2, 7]
maximum = max(numbers, key=lambda x: x)


In the above lambda function, key=lambda x: x specifies that the maximum value should be determined based on the actual values in the list.

You can use this lambda function by calling the max() function and passing the list of integers as the first argument. The lambda function is used as the key parameter to define the comparison criterion. Here's an example:

In [20]:
numbers = [5, 3, 9, 2, 7]
maximum = max(numbers, key=lambda x: x)
print(maximum) 

9


In the example above, max(numbers, key=lambda x: x) finds the maximum value in the numbers list, and it returns 9, which is the largest value in the list.

# 8. Implement a lambda function to filter out all the even numbers from a list of integers.

In [21]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
filtered = list(filter(lambda x: x % 2 == 0, numbers))


In the lambda function, x % 2 == 0 is the condition to check if x is an even number.

To use this lambda function, you can call the filter() function and pass the lambda function as the first argument. The lambda function will be used to filter the elements of the list based on the given condition. The list() function is then used to convert the filtered result into a list. Here's an example:

In [22]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
filtered = list(filter(lambda x: x % 2 == 0, numbers))
print(filtered)  


[2, 4, 6, 8, 10]


In the example above, filter(lambda x: x % 2 == 0, numbers) filters out all the even numbers from the numbers list and returns a filtered result. The resulting list, [2, 4, 6, 8, 10], contains only the even numbers from the original list.

# 9. Write a lambda function to sort a list of strings in ascending order based on the length of each string.


In [23]:
strings = ['apple', 'banana', 'cherry', 'date', 'elderberry']
sorted_strings = sorted(strings, key=lambda x: len(x))


In the lambda function, len(x) is used as the key to determine the length of each string x.

To use this lambda function, you can call the sorted() function and pass the list of strings as the first argument. The lambda function is used as the key parameter to define the sorting criterion. The resulting list will be sorted in ascending order based on the length of each string. Here's an example:

In [24]:
strings = ['apple', 'banana', 'cherry', 'date', 'elderberry']
sorted_strings = sorted(strings, key=lambda x: len(x))
print(sorted_strings)  

['date', 'apple', 'banana', 'cherry', 'elderberry']


In the example above, sorted(strings, key=lambda x: len(x)) sorts the strings list in ascending order based on the length of each string.

The resulting list, ['date', 'apple', 'cherry', 'banana', 'elderberry'], contains the strings sorted by their length from shortest to longest.

# 10. Create a lambda function that takes two lists as input and returns a new list containing the

In [25]:
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]
common_elements = list(filter(lambda x: x in list1, list2))


In the lambda function, x in list1 checks if an element x exists in list1.

To use this lambda function, you can call the filter() function and pass the lambda function as the first argument. The lambda function is used to filter the elements of list2 based on whether they are present in list1. The list() function is then used to convert the filtered result into a list. Here's an example:

In [26]:
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]
common_elements = list(filter(lambda x: x in list1, list2))
print(common_elements)  # Output: [4, 5]


[4, 5]


In the example above, filter(lambda x: x in list1, list2) filters out the elements from list2 that are present in list1 and returns a filtered result. The resulting list, [4, 5], contains the common elements between list1 and list2.

# 11. Write a recursive function to calculate the factorial of a given positive integer.

In [27]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)


In the recursive function, the base case is when n is equal to 0, in which case the factorial is defined as 1. For any other positive integer n, the function recursively calls itself with n-1 as the argument and multiplies it by n.

To use this recursive function, you can call it and pass the desired positive integer as an argument. Here's an example:

In [29]:
result = factorial(5)
print(result)  


120


In the example above, factorial(5) calls the recursive function with 5 as the input, and it returns the factorial of 5, which is 120.

# 12. Implement a recursive function to compute the nth Fibonacci number.

In [30]:
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)


In the recursive function, the base cases are when n is less than or equal to 1, in which case the Fibonacci number is equal to n itself. For any other value of n, the function recursively calls itself with n-1 and n-2 as arguments and returns the sum of the two previous Fibonacci numbers.

To use this recursive function, you can call it and pass the desired value of n as an argument. Here's an example:

In [32]:
result = fibonacci(6)
print(result)  


8


In the example above, fibonacci(6) calls the recursive function with 6 as the input, and it returns the 6th Fibonacci number, which is 8.

# 13. Create a recursive function to find the sum of all the elements in a given list.

In [33]:
def sum_list(lst):
    if not lst:
        return 0
    else:
        return lst[0] + sum_list(lst[1:])


In the recursive function, the base case is when the input list lst is empty. In that case, the sum is defined as 0. For a non-empty list, the function recursively calls itself with the sublist starting from the second element (lst[1:]) and adds it to the first element (lst[0]).

To use this recursive function, you can call it and pass the desired list as an argument. Here's an example:

In [34]:
result = sum_list([1, 2, 3, 4, 5])
print(result)


15


In the example above, sum_list([1, 2, 3, 4, 5]) calls the recursive function with the list [1, 2, 3, 4, 5] as the input, and it returns the sum of all the elements in the list, which is 15.

# 14. Write a recursive function to determine whether a given string is a palindrome.

In [35]:
def is_palindrome(string):
    if len(string) <= 1:
        return True
    else:
        return string[0] == string[-1] and is_palindrome(string[1:-1])


In the recursive function, the base case is when the length of the string is less than or equal to 1. In that case, the string is considered a palindrome and the function returns True. 

For a string with more than one character, the function checks if the first character (string[0]) is equal to the last character (string[-1]). If they are equal, the function recursively calls itself with the string excluding the first and last characters (string[1:-1]).

To use this recursive function, you can call it and pass the desired string as an argument. Here's an example:

In [36]:
result = is_palindrome("radar")
print(result) 


True


in the example above, is_palindrome("radar") calls the recursive function with the string "radar" as the input, and it returns True because "radar" is a palindrome.

# 15. Implement a recursive function to find the greatest common divisor (GCD) of two positive integers.

In [37]:
def gcd(a, b):
    if b == 0:
        return a
    else:
        return gcd(b, a % b)


In the recursive function, the base case is when the second integer b is equal to 0. In that case, the GCD is equal to the first integer a. For any other values of a and b, the function recursively calls itself with b and the remainder of the division of a by b (a % b).

To use this recursive function, you can call it and pass the two positive integers as arguments. Here's an example:

In [39]:
result = gcd(24, 36)
print(result)  


12


In the example above, gcd(24, 36) calls the recursive function with 24 and 36 as the inputs, and it returns the greatest common divisor of the two numbers, which is 12.